### ISSUE TITLE
'gcd' in variable actions evaluates to 0 and is misused (i.e., Frost DK APL)

### ISSUE THREAD

**Issue by Hekili** [simulationcraft/simc#7363]
### Describe the bug
Variables are very often used to simplify/reuse criteria in action lists, but the `gcd` expression appears to evaluate to `0` since it is a member of the `variable` action rather than the ability being tested.

### Expected behavior
I *think* folks are expecting variables to function like the old aliases(?) that were rarely used to replace criteria in the APL.  My assumption is that they are expecting `gcd` in a variable to evaluate to `gcd.max` or possibly `gcd.remains` -- though there are other places where variables divide by gcd (`%gcd`) so both `gcd` and `gcd.remains` are risks for dividing by zero.

Either:
a) APLs should be using `gcd.max` (or whatever the intended value is) instead of `gcd`, or
b) `gcd` in variable actions should evaluate to `gcd.max` (or other desired value) instead of zero.

### To Reproduce

Frost DK priority has `rp_buffs` variable that uses `3*gcd` in time comparisons:
> `variable,name=rp_buffs,value=talent.unleashed_frenzy&(buff.unleashed_frenzy.remains<gcd*3|buff.unleashed_frenzy.stack<3)|talent.icy_talons&(buff.icy_talons.remains<gcd*3|buff.icy_talons.stack<3)`

https://www.raidbots.com/simbot/report/n4mD4czULgU7ErBUhzD6dZ/simc 

Sorry for using a L60 character; I didn't have a 70 available but it would not impact this issue.  Here's an entry in the above sim that references the above variable (see Hokulea actor):

\# | n | Action + Conditions
-|-|:-
H | 1.80 | frost_strike,if=active_enemies=1&variable.rp_buffs&!talent.breath_of_sindragosa&talent.obliteration&!buff.pillar_of_frost.up

It's used less than twice on average, but clearly `rp_buffs` evaluates to true at some point.  I made two edits to the priority in the copy labeled OblitElseZeroAndRPBuffs.

1. I changed `rp_buffs` so it doesn't rely on stack counts for UF or IT; it will only be 1/true based on the buffs' remaining times.

> `variable,name=rp_buffs,value=talent.unleashed_frenzy&(buff.unleashed_frenzy.remains<gcd*3)|talent.icy_talons&(buff.icy_talons.remains<gcd*3)`

2. I added an entry for Howling Blast contingent only on `rp_buffs` being 1/true.

> `howling_blast,if=variable.rp_buffs`

You can see that this entry for Howling Blast is never used; `rp_buffs` never evaluates to 1/true because UF remains and IT remains is never less than zero.  The entry for Frost Strike that *was* used is no longer used; it was only firing due to stack count, never remaining time.

\# | n | Action + Conditions
-|-|:-
  | 0.00 | howling_blast,if=variable.rp_buffs
  | 0.00 | frost_strike,if=active_enemies=1&variable.rp_buffs&!talent.breath_of_sindragosa&talent.obliteration&!buff.pillar_of_frost.up

Relevant lines with `gcd` in variables from Frost DK:
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L205
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L208
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L209

Note that changing `gcd` to `gcd.max` isn't necessarily a complete solution.  If you do so, `pooling_runes` evaluates to 1/true from the start, so you are never able to do your initial Obliterate and just auto-attack until you have enough RP to do other things.

See GcdMax here:  https://www.raidbots.com/simbot/report/nHEtCBVciRVM5X8TXFx3Rc

Essentially, the Frost DK profile (if talented for Obliteration) relies on current issue because:
> `variable,name=oblit_pooling_time,op=setif,value=((cooldown.pillar_of_frost.remains_expected+1)%gcd)%((rune+3)*(runic_power+5))*100,value_else=gcd*2,condition=runic_power<35&rune<2&cooldown.pillar_of_frost.remains_expected<10`

...will evaluate to `value_else=gcd*2` initially, with `gcd*2` resulting in zero, then `pooling_runes` becomes 0/false here:
> `variable,name=pooling_runes,value=talent.obliteration&cooldown.pillar_of_frost.remains_expected<variable.oblit_pooling_time`

...because `cooldown.pillar_of_frost.remains_expected` (0) is not less than `variable.oblit_pooling_time` (also 0).

To address that (at least partially, maybe, depending on intent), I added `rune<6` to the `pooling_runes` variable and obtained an appropriate result (See GcdMaxRuneCheck in the same sim).

tl;dr: You probably don't want to rely on multiples of `gcd` in variables when `gcd` is always zero there.  Whether you adjust APLs or adjust `gcd`'s value in variables, additional adjustments might be needed since current logic may accidentally rely on the current behavior.

### Additional information
* SimulationCraft Version (required): git build [06a22a2ea6](https://github.com/simulationcraft/simc/commit/06a22a2ea6)

* ~~If SimulationCraft is crashing, please include your **dxdiag** file - Instructions here - 
  https://us.battle.net/support/en/article/Obtaining-System-Files~~

* ~~**simc_gui.simc** file or contents
  This file is saved in the same folder as SimulationCraft after each simulation. Make sure to run the 
  simulation that is giving you issues, then include the file here by either uploading to github, 
  copy/pasting the contents, or using http://pastebin.com/~~

**Comment by Hekili** [comment]
The only other priority where I found an impact is Unholy (`festermight_tracker`): 
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L378

**Comment by Hekili** [comment]
If you want to broaden the scope of this ticket, the same issue likely occurs for `call_action_list` and `run_action_list`.

Frost DK:
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L228

Guardian:
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/guardian_apl.inc#L21

Destruction:
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/warlock.cpp#L283
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/warlock.cpp#L306

**Comment by Pewtro** [comment]
FYI @Azevara (Lock) @nyterage (DK) @gastank (Druid)

**Comment by nyterage** [comment]
Should now be fixed for DK now, let me know if there are any further issues. 

Addressed in 
https://github.com/simulationcraft/simc/commit/be31667a570638189db1658b39a062e84b54cdfb
&
https://github.com/simulationcraft/simc/commit/658da9c4e181f3066d6ce4a71903dae0e612f6b4

**Comment by Hekili** [comment]
@nyterage I haven't built and tested yet, but I suspect that simming with Obliteration talented is going to break down because `variable.pooling_runes` is still going to be 1/true and prevent you from using Obliteration at the start of the fight since PoF is ready.  I think either `pooling_runes` needs to check that you're not capped on Runes or Pillar of Frost logic needs to change.

**Comment by nyterage** [comment]
Ah good catch, 
should address it here. Was doing my tests on PTR sims, which due to some APL changes, didnt suffer from this issue in tests. 
https://github.com/simulationcraft/simc/commit/8adc7eaaa6946f1823b676e00c0467539a7633d0

**Comment by Hekili** [comment]
@nyterage There's a second `gcd` in these Frost DK lines as well (for Icy Talons):
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L207
* https://github.com/simulationcraft/simc/blob/dragonflight/engine/class_modules/apl/apl_death_knight.cpp#L361

**Comment by nyterage** [comment]
https://github.com/simulationcraft/simc/commit/7d00d4c6ceccdfe193528204c55f333eef8abe65

should be fixed now
